engine/common/follower/cache/cache.go

```diff
@@ -0,0 +1,144 @@
+// Concurrency safe.
+type Cache struct {
+	backend *herocache.Cache // cache with random ejection
+	lock    sync.RWMutex
+	// secondary index by view, can be used to detect equivocation
+	byView map[uint64]*flow.Block
+	// secondary index by parentID, can be used to find child of the block
+	byParent map[flow.Identifier]*flow.Block
+	// when message equivocation has been detected report it using this callback
+	onEquivocation OnEquivocation
+}
+
+// Peek performs lookup of cached block by blockID.
+// Concurrency safe
+func (c *Cache) Peek(blockID flow.Identifier) *flow.Block {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+	if block, found := c.backend.ByID(blockID); found {
+		return block.(*flow.Block)
+	} else {
+		return nil
+	}
+}
+
+// NewCache creates new instance of Cache, as part of construction process connects ejection event from HeroCache to
+// post-ejection processing logic to perform cleanup of secondary indexes to prevent memory leaks.
+func NewCache(log zerolog.Logger, limit uint32, collector module.HeroCacheMetrics, onEquivocation OnEquivocation) *Cache {
+	distributor := NewDistributor(collector)
+	cache := &Cache{
+		backend: herocache.NewCache(
+			limit,
+			herocache.DefaultOversizeFactor,
+			heropool.RandomEjection,
+			log.With().Str("follower", "cache").Logger(),
+			distributor,
+		),
+		byView:         make(map[uint64]*flow.Block, 0),
+		byParent:       make(map[flow.Identifier]*flow.Block, 0),
+		onEquivocation: onEquivocation,
+	}
+	distributor.AddConsumer(cache.handleEjectedEntity)
+	return cache
+}
+
+// handleEjectedEntity performs cleanup of secondary indexes to prevent memory leaks.
+// WARNING: Concurrency safety of this function is guaranteed by s.lock, this callback can be called
+// only in herocache.Cache.Add and we perform this call while s.lock is in locked state.
+func (c *Cache) handleEjectedEntity(entity flow.Entity) {
+	block := entity.(*flow.Block)
+	delete(c.byView, block.Header.View)
+	delete(c.byParent, block.Header.ParentID)
+}
+
+// AddBlocks atomically applies batch of blocks to the cache of pending but not yet certified blocks. Upon insertion cache tries to resolve
+// incoming blocks to what is stored in the cache.
+// We require that incoming batch is sorted by height and doesn't have skipped blocks.
+// When receiving batch: [first, ..., last], we are only interested in first and last blocks since all other blocks will be certified by definition.
+// Next scenarios are possible:
+// - for first block:
+//   - no parent available for first block, we need to cache it since it will be used to certify parent when it's available.
+//   - parent for first block available in cache allowing to certify it, no need to store first block in cache.
+//
+// - for last block:
+//   - no child available for last block, we need to cache it since it's not certified yet.
+//   - child for last block available in cache allowing to certify it, no need to store last block in cache.
```

Comment:
The comments on lines 80 and 84 seem to contradict the implementation, in particular this section:
```
// store all blocks in the cache to provide deduplication
c.backend.Add(block.ID(), block)
```

Outdated comments maybe?

Reply:
updated comments: https://github.com/onflow/flow-go/pull/3965/commits/7c9a3add0b12bacd969fa30ec8a73f201ea74fc9